compiler/rustc_middle/src/traits/mod.rs

```diff
@@ -123,6 +124,14 @@ pub struct ObligationCauseData<'tcx> {
     pub code: ObligationCauseCode<'tcx>,
 }
 
+impl Hash for ObligationCauseData<'_> {
```

Comment:
For next time: I suggest leaving a brief comment explaining why this custom hash is used, because it's non-obvious and no fun to have to dig through version control history to understand why.

Reply:
Please open an issue so it's not forgotten.

Reply:
I'm not opposed to adding comments but 
> no fun to have to dig through version control history 

seems a bit subjective, for me that's 3 clicks which is hardly onerous. So that didn't quite meet my threshold of adding a comment.